### PR TITLE
prov/gni: Implement triggered ops

### PR DESCRIPTION
**Pull request by ztiffany**
-convert freelist to use dlist
-convert gnix_fab_req to use dlist
-Implement FI_TRIGGER transaction flag

Fixes ofi-cray/libfabric-cray#453.

Signed-off-by: Zach Tiffany ztiffany@cray.com

@sungeunchoi @hppritcha @chuckfossen

### CODE REVIEW

**Comment by ztiffany** [comment]
I accidentally combined the two commits I had here.  One had all the changes to the freelist.  The second commit would have only had triggered ops stuff.  I hope that's not a deal breaker.

**Comment by hppritcha** [comment]
No problem with the single commit.

File: prov/gni/include/gnix.h [hunk 1]
```diff
@@ -708,6 +711,10 @@ struct gnix_fab_req {
 	void                      *txd;
 	uint32_t                  tx_failures;
 
+	union {
+		struct fi_trigger_threshold trigger_threshold;
+	};
+
```
Comment (ztiffany) [inline prov/gni/include/gnix.h]: This wasn't needed.

**Comment by lanl-ompi** [comment]
Test FAILed.

File: prov/gni/include/gnix_ep.h [hunk 2]
```diff
@@ -143,17 +143,17 @@ typedef ssize_t (*trecvmsg_func_t)(struct fid_ep *ep,
...
 	while (ret == -FI_EAGAIN)
-		ret = _gnix_sfe_alloc(&se, &ep->fr_freelist);
+		ret = _gnix_sfe_alloc(&de, &ep->fr_freelist);
 
 	if (ret == FI_SUCCESS) {
-		fr = container_of(se, struct gnix_fab_req, slist);
+		fr = container_of(de, struct gnix_fab_req, dlist);
 		fr->gnix_ep = ep;
-		fr->slist.next = NULL;  /* slist stuff isn't too smart */
+		dlist_init(&fr->dlist);  /* dlist stuff isn't too smart */
```
Comment (hppritcha) [inline prov/gni/include/gnix_ep.h]: does the isn't too smart comment still apply?

File: prov/gni/include/gnix_freelist.h [hunk 3]
```diff
@@ -56,10 +56,10 @@ extern "C" {
  * @var growth_factor      Factor for increasing refill size
  * @var max_refill_size;   Max refill size
  * @var elem_size          Size of element (in bytes)
- * @var offset             Offset of slist_entry field (in bytes)
+ * @var offset             Offset of dlist_entry field (in bytes)
  */
 struct gnix_s_freelist {
-	struct slist freelist;
+	struct dlist_entry freelist;
```
Comment (hppritcha) [inline prov/gni/include/gnix_freelist.h:62]: should the struct be renamed to `gnix_d_freelist`?

**Comment by hppritcha** [comment]
bot:retest
I can't figure what the problem is with /tmp on these cori login nodes.

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by hppritcha** [comment]
bot:retest

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by hppritcha** [comment]
bot:retest

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by hppritcha** [comment]
okay once more
bot:retest

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by hppritcha** [comment]
bot:retest

File: prov/gni/include/gnix_ep.h (hunk 2 above)
Comment (chuckfossen) [inline prov/gni/include/gnix_ep.h]: I don't think the comment applies anymore.

File: prov/gni/test/rdm_dgram_rma.c [hunk 4]
```diff
@@ -1622,3 +1622,69 @@ Test(dgram_rma, write_alignment_retrans)
...
+	msg[0].desc = (void **)loc_mr;
+	msg[0].iov_count = 1;
+	msg[0].addr = gni_addr[1];
+	msg[0].rma_iov = &rma_iov;
+	msg[0].rma_iov_count = 1;
+	msg[0].data = (uint64_t)target;
+	msg[1] = msg[2] = msg[3] = msg[0];
+
+	/* XXX: Req 0 is guaranteed to be sent before req 2, but req 2 will
+	 * race req 1 through the network.  Fix race if needed. */
```
Comment (ztiffany) [inline prov/gni/test/rdm_dgram_rma.c]: "req 1" should be "req 0"

**Comment by chuckfossen** [comment]
I mostly reviewed the changes for the ops_allowed cleanup.
:+1:

**Comment by sungeunchoi** [comment]
For my own info, what enabled the moving of `_gnix_nic_tx_free` to a single location?

**Comment by ztiffany** [comment]
When a request must fail due to insufficient capabilities it is freed before any transaction is done.  Requests were freed (in the case of RMA) in __gnix_rma_fr_complete().  Previously, this function took a request and TXD pointer.  We don't actually have a TXD in the case where we never initiated the transaction, though.  So, I removed the TXD parameter to __gnix_rma_fr_complete() to make it more general.  It just turned out that the change resulted in less code.  I think having the TXD parameter to __gnix_rma_fr_complete() was a relic from when I had the idea to reuse the TXD assigned to a request in cases where a single request required multiple, serial transfers.

**Comment by sungeunchoi** [comment]
I'm good with this, but I think we're waiting on @hppritcha

**Comment by hppritcha** [comment]
i'm good with this.  please merge.

**Comment by hppritcha** [comment]
please go ahead and merge.

**Pull request merged**